clippy_lints/src/regex.rs

```diff
@@ -55,6 +56,44 @@ declare_clippy_lint! {
     "trivial regular expressions"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Checks for [regex](https://crates.io/crates/regex) compilation inside a loop with a literal.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Compiling a regex is a much more expensive operation than using one, and a compiled regex can be used multiple times.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// # let haystacks = [""];
+    /// # const MY_REGEX: &str = "a.b";
+    /// for haystack in haystacks {
+    ///     let regex = regex::Regex::new(MY_REGEX).unwrap();
+    ///     if regex.is_match(haystack) {
+    ///         // Perform operation
+    ///     }
+    /// }
+    /// ```
+    /// should be replaced with
+    /// ```no_run
+    /// # let haystacks = [""];
+    /// # const MY_REGEX: &str = "a.b";
+    /// let regex = regex::Regex::new(MY_REGEX).unwrap();
```

Comment:
I could imagine a situation where the regex creation is in an unlikely (error) path in a loop and the suggested change of simply moving the `Regex::new()` call outside the loop would go from compiling it almost never to always compiling it once

Though in those cases one can still move it outside the loop (or even into a `static`) wrapped in a Lazy{Cell,Lock} so it's only compiled when accessed. The lint messages/description doesn't contradict this or specify/require how it should be moved out of the loop, but I think it'd still be useful to mention that somewhere because it might be non-obvious that that's an option

Reply:
It would also be good to link to regex's own description of the antipattern: https://docs.rs/regex/latest/regex/#avoid-re-compiling-regexes-especially-in-a-loop

Reply:
Done, reworded to "can be replaced with" instead of "should" and added link to docs.